### PR TITLE
OIR Updates Required for Optimizations

### PR DESCRIPTION
**Pull request by fthaler**
## Description

Grouping of vertical loops and cache declarations on the OIR level, as required for optimizations.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: src/gtc/oir.py [hunk 1]
```diff
@@ -145,19 +145,68 @@ class Interval(LocNode):
     start: AxisBound
     end: AxisBound
 
+    @root_validator
+    def check(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        start, end = values["start"], values["end"]
+        if start.level == common.LevelMarker.END and end.level == common.LevelMarker.START:
+            raise ValueError("Start level must be smaller or equal end level")
+        if start.level == end.level and start.offset > end.offset:
+            raise ValueError("Start offset must be smaller or equal end offset")
```
Comment (havogt) [inline src/gtc/oir.py:154]: I don't know which is correct, but text and condition don't match `>=` or `>`?
Also maybe add "If start and end level are equal, start offset..."

File: src/gtc/oir.py [hunk 2]
```diff
@@ -145,19 +145,68 @@ class Interval(LocNode):
...
+        return values
 
-class VerticalLoop(LocNode):
+
+class CacheDecl(LocNode):
```
Comment (havogt) [inline src/gtc/oir.py:158]: Maybe 
```suggestion
class CacheDesc(LocNode):
```
to distinguish it from the other `Decl`s which are semantically different.

**Review by havogt — CHANGES_REQUESTED** [review]
Looks good, 2 small things.

File: src/gtc/oir.py (hunk 1 above)
Comment (fthaler) [reply src/gtc/oir.py:154]: I think it does match (`start.offset > end.offset` as used in the error condition is the negation of the required condition `start.offset <= end.offset`). But maybe `not (start.offset <= end.offset)` would be more readable?

Comment (gronerl) [reply src/gtc/oir.py:154]: Interval ranges are exclusive, so shouldn't it be
 `start.level == end.level and start.offset >= end.offset` (or `not (start.offset < end.offset)`)
 Except if we want to allow empty intervals to be represented.

Comment (fthaler) [reply src/gtc/oir.py]: No idea if we need empty intervals, I just cared about ordered start and end. I disabled empty intervals now, let’s see if the tests pass…

**Review by havogt — APPROVED** [review]

**Pull request merged**